ExpressLRS/ExpressLRS
# Add telemetry support

This PR adds (optional) telemetry support.

Three new components are added:

- Telemetry Receiver
- Stubborn Sender
- Stubborn Receiver

### Telemetry Receiver

This is a rewrite of the crsf telemetry receiver that was already implemented. The receiver buffers the last received telemetry messages and has support to get the next updated messages. The following telemetry messages are supported: GPS, BATTERY_SENSOR, ATTITUDE, DEVICE_INFO, FLIGHT_MODE

Even if the telemetry support is disabled, some parts are required since the bootloader for stm32 targets is called via telemetry messages.

### Stubborn Sender
The sender runs on the receiver on the quad. In HandleSendTelemetryResponse two different types of telemetry packages are sent to the tx: ELRS_TELEMETRY_TYPE_LINK (as before) and ELRS_TELEMETRY_TYPE_DATA. The new message sends 5 bytes at the same time until the tx flips one bit to confirm the reception. Then the next 5 bytes are sent until the bit is flipped again. This simple mechanism ensures that the data isn't duplicated and that the retransmit happens until the reception is confirmed.

If sender/receiver lose sync there is an automatic sync recovery in place.

ELRS_TELEMETRY_TYPE_LINK and ELRS_TELEMETRY_TYPE_DATA is sent in an alternating order. 

- To get the same link update rate the telemetry rate can be set to 32. 
- If you set a telemetry rate of 16, the ELRS_TELEMETRY_TYPE_DATA is sent twice as often. This way the link update happens at the same frequency as in the case of the telemetry rate of 64 but the telemetry data is sent faster

### Stubborn Receiver
The receiver runs on the tx. The confirmation bit is sent on the slot of one of the 1-Bit switches or on the free bit in the hybrid switch data. When the receiver receives one complete frame, it's sent to opentx.

There are native tests for all modules that can be run without any hardware. I extracted some generic crsf definitions to a separte header file so I can use this defines without pulling in hardware dependent includes.

### Test the PR

- Enable telemetry in your FC
- Scan for sensors on TX
- check if you see the voltage RxBt value on the TX to verify that basic telemetry is working
- Download my fork of the repository: https://github.com/schugabe/ExpressLRS 
- Open it in Visual Studio Code
- In bottom left of the IDE you see "master-dev", click on it and select telemetry
- Edit user settings: make sure you have the lines with "-DENABLE_TELEMETRY" and "-DHYBRID_SWITCHES_8" enabled
- Flash TX + RX
- Scan for sensors the list should be longer
- Increase the telemetry rate with the expresslrs lua script. Increase the rate until the sensor lost warnings go away. I use up to 1:16 with 200Hz
